routing/DELEGATED_ROUTING_HTTP.md

```diff
@@ -0,0 +1,123 @@
+
+"Delegated routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated routing.
+
+# Organization of this document
+
+- [Introduction](#introduction)
+- [Spec](#spec)
+  - [Interaction Pattern](#interaction-pattern)
+  - [Cachability](#cachability)
+  - [Transports](#transports)
+  - [Protocol Message Overview](#protocol-message-overview)
+    - [Known Methods](#known-methods)
+- [Method Upgrade Paths](#method-upgrade-paths)
+- [Implementations](#implementations)
+
+# API Specification
+The Delegated Routing HTTP API uses the `application/json` content type by default. Clients and servers *should* support `application/cbor`, which can be negotiated using the standard `Accept` and `Content-Type` headers.
+
+## Common Data Types:
+
+- CIDs are always encoded using a [multibase](https://github.com/multiformats/multibase)-encoded [CIDv1](https://github.com/multiformats/cid#cidv1).
+- Multiaddrs are encoded according to the [human-readable multiaddr specification](https://github.com/multiformats/multiaddr#specification)
+- Peer IDs are encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use Base64.
+
+## API
+- `GET /v1/providers/{CID}`
+    - Reframe equivalent: FindProviders
+    - Response
+        
+        ```json
+        {
+            "Providers": [
+                {
+                    "PeerID": "12D3K...",
+                    "Multiaddrs": ["/ip4/.../tcp/.../p2p/...", "/ip4/..."],
+                    "Protocols": [
+                        {
+                            "Codec": 2320,
+                            "Payload": { ... }
+                        }
+                    ]
+                }
+            ]
+        }
+        ```
+        
+    - Default limit: 100 providers
+    - Optional query parameters
+        - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of [multicodec codes](https://github.com/multiformats/multicodec/blob/master/table.csv) in decimal form such as `2304,2320`
+        - `transport` only return providers whose published multiaddrs explicitly support the passed transport protocols, such as `460,478` (`/quic` and `/tls/ws`)
+        - Servers should treat the multicodec codes used in the `transfer` and `transport` parameters as opaque, and not validate them, for forwards compatibility
+- `GET /v1/providers/hashed/{multihash}`
+    - This is the same as `GET /v1/providers/{CID}`, but takes a hashed CID encoded as a [multihash](https://github.com/multiformats/multihash/)
+- `GET /v1/ipns/{ID}`
+    - Reframe equivalent: GetIPNS
+    - `ID`: multibase-encoded bytes
+    - Response
+        - record bytes
+- `POST /v1/ipns`
```

Comment:
Reframe does not validate IPNS records like it does provider records.  This seems like a miss, shouldn't the server be validating these records before accepting them?

Reply:
I don't think it is the responsibility of HTTP Client / Server library – baloons the complexity.
IPNS validation should happen later, after record is passed to business logic.

Reply:
My main concern is that this is inconsistent w/ provider records, are you saying we shouldn't validate provider records either? Or is there something different w/ IPNS that doesn't justify keeping them consistent?

Reply:
I have remove IPNS from this spec and rename it to "delegated content routing", so this is now moot